---
title: "dbsnp_hg38"
author: "Lily Andrews"
date: "2023-07-03"
output: "github_document"
---
## This code shows you how to download dbSNP data for hg38 alignment, which you can use to map rsID to chromosome and position or vice versa.
```{bash}
#download the dbSNP data aligned to hg38
#"-O" specifies file name
wget -O dbsnp_hg38.vcf.gz http://fileserve.mrcieu.ac.uk/dbsnp/dbsnp.v153.hg38.vcf.gz 
wget -O dbsnp_hg38.vcf.gz.tbi http://fileserve.mrcieu.ac.uk/dbsnp/dbsnp.v153.hg38.vcf.gz.tbi 
#create a file is created to rename the chromosomes so that harmonization to other datasets is easier
echo -n "NC_000001.11    1
NC_000002.12    2
NC_000003.12    3
NC_000004.12    4
NC_000005.10    5
NC_000006.12    6
NC_000007.14    7
NC_000008.11    8
NC_000009.12    9
NC_000010.11    10
NC_000011.10    11
NC_000012.12    12
NC_000013.11    13
NC_000014.9     14
NC_000015.10    15
NC_000016.10    16
NC_000017.11    17
NC_000018.10    18
NC_000019.10    19
NC_000020.11    20
NC_000021.9     21
NC_000022.11    22
NC_000023.11    X
NC_000024.10    Y
NC_012920.1     MT
" > hg38_rename_chrom_names.tsv
#this code changes dbSNP chromosome ID to more harmonisable chromosome name 
bcftools annotate --rename-chrs hg38_rename_chrom_names.tsv --output-type z --output dbsnp_hg38_clean.vcf.gz dbsnp_hg38.vcf.gz 
#index new file
bcftools index dbsnp_hg38_clean.vcf.gz
```

## Here's an example of how to map chromsome and position of UKBB pQTL data to rsID of dbSNP
```{r}
#load in ukbb protein data
prot <- read.table("ukbb_protein.txt", header=T) 
marker <- prot$ID
#split marker name to chromosome position and alleles
position <- data.frame(do.call("rbind", strsplit(as.character(marker), ":", fixed = TRUE))) 
#take chromosome and position
subset <- position[, 1:2] 
#add chr to chromosome list as this with match dbsnp file
subset$X1<- sub("^", "chr", subset$X1) 
#add chromosome hg38 to original dataframe
prot$CHROM <- subset$X1
#add position hg38 to original dataframe
prot$GENPOS <- subset$X2 
#hg38 alignment 
vcf <- "dbsnp_hg38_clean.vcf.gz" 
#create temporary file
tmp <- tempfile() 
#"--region-file" runs better with 3 columns than 2, so the position column is duplicated
subset$X3<- subset$X2
#no row names are needed
rownames(subset) <- NULL
#no column names are needed
colnames(subset) <- NULL 
#create a file for subset data
write.table(subset, file=paste0("region_file.txt"), row.names=FALSE, col.names=FALSE, quote=FALSE,sep="\t") 
#path to bcftools
gwasvcf::set_bcftools("bcftools") 
bcftools <- "bcftools" 
#from chromosome and position to rsid
cmd <- glue("{bcftools} query -f '%CHROM %POS %ID\n' {vcf} --regions-file region_file.txt > {tmp}.txt") 
#run this on the command line of linux operating system
system(cmd) 
#mapped file created
map <- fread(glue("{tmp}.txt")) 
colnames(map) <- c("CHROM", "GENPOS", "SNP")
#map rsids into original protein dataframe
prot <- merge(prot, map, by="GENPOS") 
#create output of protein dataframe now with rsid from chromosome and position
write.table(prot, file=paste0("mapped_", x, ".txt")) 
```

