---
title: "download_dbsnp_hg38"
author: "Lily Andrews"
date: "2023-07-03"
output: "github_document"
---
##This code shows you how to download dbSNP data for hg38 alignment, which you can use to map rsID to chromosome and position or vice versa.
```{bash}
wget http://fileserve.mrcieu.ac.uk/dbsnp/dbsnp.v153.hg38.vcf.gz 
wget http://fileserve.mrcieu.ac.uk/dbsnp/dbsnp.v153.hg38.vcf.gz.tbi 
echo -n "NC_000001.11    1
NC_000002.12    2
NC_000003.12    3
NC_000004.12    4
NC_000005.10    5
NC_000006.12    6
NC_000007.14    7
NC_000008.11    8
NC_000009.12    9
NC_000010.11    10
NC_000011.10    11
NC_000012.12    12
NC_000013.11    13
NC_000014.9     14
NC_000015.10    15
NC_000016.10    16
NC_000017.11    17
NC_000018.10    18
NC_000019.10    19
NC_000020.11    20
NC_000021.9     21
NC_000022.11    22
NC_000023.11    X
NC_000024.10    Y
NC_012920.1     MT
" > hg38_rename_chrom_names.tsv #a file is created to rename the chromosomes so that harmonisation to other datasets is easier.
bcftools annotate --rename-chrs hg38_rename_chrom_names.tsv --output-type z --output dbSNP_clean.vcf.gz dbsnp.v153.hg38.vcf.gz #this code changes dbSNP chromsome ID to more harmonisable chromsome name 
bcftools index dbSNP_clean.vcf.gz
```
## Here's an example of how to map chromsome and position of UKBB pQTL data to rsID of dbSNP
```{r}
prot <- read.table("ukbb_protein.txt", header=T) #load in ukbb protein data
marker <- prot$ID
position <- data.frame(do.call("rbind", strsplit(as.character(marker), ":", fixed = TRUE))) #split marker name to chromsome position and alleles
subset <- position[, 1:2] #take chromosome and position
subset$X1<- sub("^", "chr", subset$X1) #add chr to chromosome list as this with match dbsnp file
prot$CHROM <- subset$X1 #add chromosome hg38 to original dataframe
prot$GENPOS <- subset$X2 #add position hg38 to original dataframe
vcf <- "dbSNP_clean.vcf.gz" #hg38 alignment 
tmp <- tempfile() #create temporary file
subset$X3<- subset$X2 #--region-file runs better with 3 columns than 2, so the position column is duplicated
rownames(subset) <- NULL #no row names are needed
colnames(subset) <- NULL #no column names are needed
write.table(subset, file=paste0("region_file.txt"), row.names=FALSE, col.names=FALSE, quote=FALSE,sep="\t") #create a file for subsetted data
gwasvcf::set_bcftools("bcftools") #path to bcftools
bcftools <- "bcftools" #path to bcftools
cmd <- glue("{bcftools} query -f '%CHROM %POS %ID\n' {vcf} --regions-file region_file.txt > {tmp}.txt") #from chromosome and position to rsid
system(cmd) #run this on the command line of linux operating system
map <- fread(glue("{tmp}.txt")) #mapped file created
colnames(map) <- c("CHROM", "GENPOS", "SNP")
prot <- merge(prot, map, by="GENPOS") #map rsids into original protein dataframe
write.table(prot, file=paste0("mapped_", x, ".txt")) #create output of protein dataframe now with rsid from chromosome and position
```

