---
title: "reference_panel_hg19"
author: "Lily Andrews"
date: "2023-07-03"
output: "github_document"
---
##This code downloads the 1000 genome data from the hg19 release in Plink
##Ensure plink is installed for the following code
```{bash} 
#Load plink 2.0.0
#"-O" specifies file name
wget -O all_phase3.psam "https://www.dropbox.com/s/6ppo144ikdzery5/phase3_corrected.psam?dl=1"
wget -O all_phase3.pgen.zst "https://www.dropbox.com/s/y6ytfoybz48dc0u/all_phase3.pgen.zst?dl=1"
wget -O all_phase3.pvar.zst "https://www.dropbox.com/s/odlexvo8fummcvt/all_phase3.pvar.zst?dl=1"
#decompress file 
plink2 --zst-decompress all_phase3.pgen.zst > all_phase3.pgen
#filter for european if required
awk 'NR == 1 || $5 == "EUR" {print $1}' all_phase3.psam > EUR_1kg_samples.txt
# "vzs" modifier to directly operate with pvar.zst
# "--output-chr 26" uses numeric chromosome codes
# "--max-alleles 2": PLINK 1 binary does not allow multi-allelic variants
#"--rm-dup" remove duplicates
#"--chr 1-22" select chromosomes and not include sex chromosomes
#"--keep" only includes the list of IDs, this is set to european but can be changed
#also make binary files
plink2 --pfile all_phase3 vzs --chr 1-22 --max-alleles 2  --output-chr 26 --rm-dup exclude-mismatch --keep EUR_1kg_samples.txt --make-bed --out all_phase3_new
#Load plink 1.90 
#create file with "." to remove SNPs (remove.txt) which are only "." which are present in this file
echo "." > remove.txt
plink --bfile all_phase3_new --extract remove.txt --make-bed --out all_phase3_new.clean
```


```{bash}
#download the dbSNP data aligned to hg19, this can be used to map chromosome and position to rsid
#"-O" specifies file name
wget -O dbsnp_hg19.vcf.gz http://fileserve.mrcieu.ac.uk/dbsnp/dbsnp.v153.b37.vcf.gz 
wget -O dbsnp_hg19.vcf.gz.tbi http://fileserve.mrcieu.ac.uk/dbsnp/dbsnp.v153.b37.vcf.gz.tbi 
#index vcf file
bcftools index dbsnp_hg19.vcf.gz
```

## Here's an example of how to map chromsome and position of UKBB pQTL data to rsID of dbSNP
```{r}
#load in ukbb protein data
prot <- read.table("ukbb_protein.txt", header=T) 
marker <- prot$ID
#split marker name to chromosome position and alleles
position <- data.frame(do.call("rbind", strsplit(as.character(marker), ":", fixed = TRUE))) 
#take chromosome and position
subset <- position[, 1:2] 
#add chr to chromosome list as this with match dbsnp file
subset$X1<- sub("^", "chr", subset$X1) 
#add chromosome hg19 to original dataframe
prot$CHROM <- subset$X1
#add position hg19 to original dataframe
prot$GENPOS <- subset$X2 
#hg38 alignment 
vcf <- "ddbsnp_hg19.vcf.gz" 
#create temporary file
tmp <- tempfile() 
#"--region-file" runs better with 3 columns than 2, so the position column is duplicated
subset$X3<- subset$X2
#no row names are needed
rownames(subset) <- NULL
#no column names are needed
colnames(subset) <- NULL 
#create a file for subset data
write.table(subset, file=paste0("region_file.txt"), row.names=FALSE, col.names=FALSE, quote=FALSE,sep="\t") 
#path to bcftools
gwasvcf::set_bcftools("bcftools") 
bcftools <- "bcftools" 
#from chromosome and position to rsid
cmd <- glue("{bcftools} query -f '%CHROM %POS %ID\n' {vcf} --regions-file region_file.txt > {tmp}.txt") 
#run this on the command line of linux operating system
system(cmd) 
#mapped file created
map <- fread(glue("{tmp}.txt")) 
colnames(map) <- c("CHROM", "GENPOS", "SNP")
#map rsids into original protein dataframe
prot <- merge(prot, map, by="GENPOS") 
#create output of protein dataframe now with rsid from chromosome and position
write.table(prot, file=paste0("mapped_", x, ".txt")) 
```


